File: executioncontext/workspacecontext/src/main/java/org/jetbrains/bsp/bazel/workspacecontext/WorkspaceContext.java
package org.jetbrains.bsp.bazel.workspacecontext;

import io.vavr.control.Option;
import io.vavr.control.Try;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.jetbrains.bsp.bazel.executioncontext.api.ExecutionContext;
import org.jetbrains.bsp.bazel.workspacecontext.entries.ExecutionContextTargetsEntity;

/**
 * Representation of <code>ExecutionContext</code> used during server lifetime.
 *
 * @see org.jetbrains.bsp.bazel.executioncontext.api.ExecutionContext
 */
public class WorkspaceContext extends ExecutionContext {

  /**
   * Targets (included and excluded) on which the user wants to work.
   *
   * <p>Obtained from <code>ProjectView</code> simply by mapping 'targets' section.
   */
  private final ExecutionContextTargetsEntity targets;

  private WorkspaceContext(ExecutionContextTargetsEntity targets) {
    this.targets = targets;
  }

  public ExecutionContextTargetsEntity getTargets() {
    return targets;
  }

  public static WorkspaceContext.Builder builder() {
    return new Builder();
  }

  public static class Builder {

    private static final Logger log = LogManager.getLogger(WorkspaceContext.Builder.class);

    private Option<ExecutionContextTargetsEntity> targets = Option.none();

    private Builder() {}

    public Builder targets(ExecutionContextTargetsEntity targets) {
      this.targets = Option.of(targets);

      return this;
    }

    public Try<WorkspaceContext> build() {
      if (targets.isEmpty()) {
        var exceptionMessage = "Workspace context creation failed! 'targets' has to be defined.";

        log.error(exceptionMessage);

        return Try.failure(new IllegalStateException(exceptionMessage));
      }

      return Try.success(new WorkspaceContext(targets.get()));
    }
  }
}

File: executioncontext/workspacecontext/src/main/java/org/jetbrains/bsp/bazel/workspacecontext/WorkspaceContext.kt
package org.jetbrains.bsp.bazel.workspacecontext

import org.apache.logging.log4j.LogManager
import org.jetbrains.bsp.bazel.executioncontext.api.ExecutionContext
import org.jetbrains.bsp.bazel.executioncontext.api.ExecutionContextConstructor
import org.jetbrains.bsp.bazel.projectview.model.ProjectView

/**
 * Representation of `ExecutionContext` used during server lifetime.
 *
 * @see org.jetbrains.bsp.bazel.executioncontext.api.ExecutionContext
 */
data class WorkspaceContext(
    /**
     * Targets (included and excluded) on which the user wants to work.
     *
     *
     * Obtained from `ProjectView` simply by mapping 'targets' section.
     */
    val targets: TargetsSpec,

    /**
     * Build flags which should be added to each bazel call.
     *
     * Obtained from `ProjectView` simply by mapping `build_flags` section.
     */
    val buildFlags: BuildFlagsSpec,

    /**
     * Path to bazel which should be used in the bazel runner.
     *
     * Obtained from `ProjectView` if not null, otherwise deducted from `PATH`.
     */
    val bazelBinary: BazelBinarySpec,

    /**
     * If true targets with `manual` tag will be built
     *
     * Obtained from `ProjectView` simply by mapping 'build_manual_targets' section.
     */
    val buildManualTargets: BuildManualTargetsSpec,

    // TODO replace `BspInfo`
    /**
     * Path to the `.bazelbsp` dir in the project root
     *
     * Deducted from working directory.
     */
    val dotBazelBspDirPath: DotBazelBspDirPathSpec,

    /**
     * Parameter determining targets importing depth
     *
     * Obtained from `ProjectView` simply by mapping `import_depth` section.
     */
    val importDepth: ImportDepthSpec,
) : ExecutionContext()


object WorkspaceContextConstructor : ExecutionContextConstructor<WorkspaceContext> {

    private val log = LogManager.getLogger(WorkspaceContextConstructor::class.java)

    override fun construct(projectView: ProjectView): WorkspaceContext {
        log.info("Constructing workspace context for: {}.", projectView)

        // maybe TRY is not that good xd
        val targetsSpec = TargetsSpecExtractor.fromProjectView(projectView)
        val buildFlagsSpec = BuildFlagsSpecExtractor.fromProjectView(projectView)
        val bazelBinarySpec = BazelBinarySpecExtractor.fromProjectView(projectView)
        val dotBazelBspDirPathSpec = DotBazelBspDirPathSpecExtractor.fromProjectView(projectView)
        val buildManualTargetsSpec = BuildManualTargetsSpecExtractor.fromProjectView(projectView)
        val importDepthSpec = ImportDepthSpecExtractor.fromProjectView(projectView)
        return WorkspaceContext(
            targets = targetsSpec,
            buildFlags = buildFlagsSpec,
            bazelBinary = bazelBinarySpec,
            dotBazelBspDirPath = dotBazelBspDirPathSpec,
            buildManualTargets = buildManualTargetsSpec,
            importDepth = importDepthSpec,
        )
    }


    override fun constructDefault(): WorkspaceContext {
        val targetsSpec = TargetsSpecExtractor.default()
        val buildFlagsSpec = BuildFlagsSpecExtractor.default()
        val bazelBinarySpec = BazelBinarySpecExtractor.default()
        val dotBazelBspDirPathSpec = DotBazelBspDirPathSpecExtractor.default()
        val buildManualTargetsSpec = BuildManualTargetsSpecExtractor.default()
        val importDepthSpec = ImportDepthSpecExtractor.default()
        return WorkspaceContext(
            targets = targetsSpec,
            buildFlags = buildFlagsSpec,
            bazelBinary = bazelBinarySpec,
            dotBazelBspDirPath = dotBazelBspDirPathSpec,
            buildManualTargets = buildManualTargetsSpec,
            importDepth = importDepthSpec,
        )
    }
}
